### ISSUE TITLE
Track when a dict originated from a literal and be more forgiving about `**kwargs` unpacking to optional parameters

### ISSUE THREAD

**Issue by rzwq** [astral-sh/ty#3792]
### Question

I'm not sure, that its bug

<img width="1584" height="270" alt="Image" src="https://github.com/user-attachments/assets/e75948d0-0173-4c0a-9344-7d12843f473d" />

<img width="1532" height="300" alt="Image" src="https://github.com/user-attachments/assets/73b9cb8e-4843-4e9c-b6e9-d5e95becae28" />

### Version

0.0.48

**Comment by sharkdp** [comment]
Code here:
```py
params = {
    "sep": "\n",
    "end": "\n\n"
}

print(*range(5), **params)
```

Works at runtime, ty/mypy/pyright all reject it, pyrefly allows it. I guess the problem for ty is that `params` is inferred as `dict[str, str]`?

**Comment by carljm** [comment]
Yeah, that's right. We track the type of `params` as `dict[str, str]`, as well as tracking the two known keys `"sep"` and `"end"` and their types -- but we don't assume those are the only two keys. So the call to `print` fails because `print` has keyword arguments `flush` and `file` which do not accept strings.

I think there is a strong argument here that in the same way `*args` unpacking with a list assumes the list is of the right length (since length is not a part of the list type), `**kwargs` unpacking with a non-TypedDict dict type (which does not model presence/absence of specific keys) should assume any keys which would map to an optional argument and which would not have the correct type, are not present in the source dict. This would be a more general change to the behavior of `**kwargs` unpacking, which would diverge from pyright/mypy. But it does appear to be the behavior of pyrefly. It would mean this example would not error, because we would optimistically assume the `"num"` key is not present in the source dict:

```py
params = {
    "sep": "\n",
    "end": "\n\n"
}

def f(sep: str, end: str, num: int = 1): ...

f(**params)
```

**Comment by carljm** [comment]
Pyrefly's behavior is actually not a generalized one; it seems to actually track an implicit internal TypedDict-like type for dict literals (which is easily rendered incorrect by even obvious mutations of the dict: add a `params["file"] = "foo"` to the above example and it will fail at runtime, but Pyrefly will still allow it).

No type checker allows this:

```py
def f(sep: str, end: str, num: int = 1): ...

def _(params: dict[str, str]):
    f(**params)
```

Which is actually parallel to the behavior of `*args` unpacking, since no type checker allows this either:

```py
args: list[str] = ["a", "b"]

def f(a: str, b: str, num: int = 1): ...

f(*args)
```

And the lax behavior I suggested above would also be problematic because it would allow any `dict[str, T]` to be unpacked into a call to any function taking only optional parameters, which is perhaps too lax.

So I think I retract the above suggestion. I'm not sure we need to make any change here, but if we did it would probably be shaped something like "track when we think a dict has only the keys we've seen added to it, and take that into account in unpacking."

**Comment by AlexWaygood** [comment]
Another possible change is, of course, "somehow improve the error message so that it's clearer to the user what we're complaining about and why"

**Comment by sharkdp** [comment]
Related problem here with construction of Pydantic models: https://github.com/astral-sh/ty/issues/4072
